File: app/src/main/java/com/digicoffer/lauditor/FirmProfile/PracticePartnerAdd.java
package com.digicoffer.lauditor.FirmProfile;

import android.app.Dialog;
import android.content.res.Resources;
import android.os.Bundle;
import android.text.Editable;
import android.text.InputType;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.activity.OnBackPressedCallback;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.LinearLayoutCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;

import com.digicoffer.lauditor.CommonFiles.GlobalFiles.NewModel;
import com.digicoffer.lauditor.R;
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener;
import com.digicoffer.lauditor.Webservice.HttpResultDo;
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.DynamicUtils;
import com.digicoffer.lauditor.CommonFiles.ValidationUtils.Validation;
import com.google.android.material.textfield.TextInputEditText;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.Objects;

public class PracticePartnerAdd extends Fragment implements AsyncTaskCompleteListener, View.OnClickListener {

    TextView tv_add_pp, tv_Fname, tv_Lname, tv_designation, tv_specialist, tv_email, tv_phonenumber;
    TextInputEditText et_Fname, et_Lname, et_designation, et_specialist, et_email, et_phonenumber;
    Button btn_cancel_pp, btn_save_pp;
    Dialog progress_dialog;
    MemberProfileModel memberProfileModel;
    FirmProfile firmProfile;
    boolean isemailhasError = false;
    ImageView iv_cancel;
    boolean isphonehasError = false;
    TextView response_email, tvErrorPhone;
    LinearLayoutCompat last_name_layout;
    private NewModel mViewModel;

    public PracticePartnerAdd(MemberProfileModel model, FirmProfile firmProfile) {
        memberProfileModel = model;
        this.firmProfile = firmProfile;
    }

    @Override
    public void onClick(View view) {
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View v = inflater.inflate(R.layout.activity_practice_partner_add, container, false);
        return v;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        setupOnBackPressed();
        super.onCreate(savedInstanceState);
    }

    private void setupOnBackPressed() {
        requireActivity().getOnBackPressedDispatcher().addCallback(new OnBackPressedCallback(true) {
            @Override
            public void handleOnBackPressed() {
                if (isEnabled()) {
//                    AndroidUtils.showToast("Dashboard",getContext());
                    setEnabled(false);
                    requireActivity().onBackPressed();
                }
            }
        });
    }

    @Override
    public void onViewCreated(@NonNull View v, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(v, savedInstanceState);
        try {
            mViewModel = new ViewModelProvider(requireActivity()).get(NewModel.class);
            tv_add_pp = v.findViewById(R.id.tv_add_pp);
            tv_Fname = v.findViewById(R.id.tv_Fname);
            tv_Lname = v.findViewById(R.id.tv_Lname);
            iv_cancel = v. findViewById(R.id.iv_cancel);
            tv_designation = v.findViewById(R.id.tv_designation);
            tv_specialist = v.findViewById(R.id.tv_specialist);
            tv_email = v.findViewById(R.id.tv_email);
            tv_phonenumber = v.findViewById(R.id.tv_phonenumber);
            last_name_layout = v.findViewById(R.id.last_name_layout);
            last_name_layout.setVisibility(View.GONE);
            et_Fname = v.findViewById(R.id.et_Fname);
            et_Fname.addTextChangedListener(new Validation(et_Fname));

            et_Lname = v.findViewById(R.id.et_Lname);
            et_Lname.setVisibility(View.GONE);

            et_designation = v.findViewById(R.id.et_designation);
            et_designation.addTextChangedListener(new Validation(et_designation));
            et_specialist = v.findViewById(R.id.et_specialist);
            et_specialist.addTextChangedListener(new Validation(et_specialist));
            response_email = v.findViewById(R.id.response_email);
            tvErrorPhone = v.findViewById(R.id.tvErrorPhone);
            et_email = v.findViewById(R.id.et_email);
            et_email.addTextChangedListener(new Validation(et_email));
            et_email.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                    response_email.setVisibility(View.GONE);
                    isemailhasError = false;
                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                }

                @Override
                public void afterTextChanged(Editable s) {
                    if (!et_email.getText().toString().trim().isEmpty()) {
                        if (!AndroidUtils.isValidEmail(et_email.getText().toString())) {
//                    tv_email.setError("Enter a valid email address");
//                    AndroidUtils.showAlert("Please enter the Email", getContext());
                            response_email.setVisibility(View.VISIBLE);
                            response_email.setText(R.string.enter_a_valid_email_address);
                            isemailhasError = true;
//                hasErrors = true;
                        } else {
                            response_email.setVisibility(View.GONE);
                            isemailhasError = false;
                        }
                    }

                }
            });
            et_phonenumber = v.findViewById(R.id.et_phonenumber);
            et_phonenumber.setInputType(InputType.TYPE_CLASS_NUMBER);
            et_phonenumber.addTextChangedListener(new Validation(et_phonenumber));
            AndroidUtils.NumberFilter(et_phonenumber, true);

            et_phonenumber.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {
                    tvErrorPhone.setVisibility(View.GONE);
                    isphonehasError = false;
                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {
                }

                @Override
                public void afterTextChanged(Editable s) {
                    if (!s.toString().isEmpty()) {
                        if (s.length() < 10) {
                            tvErrorPhone.setVisibility(View.VISIBLE);
                            tvErrorPhone.setText(R.string.please_enter_the_10_digit_mobile_number);
                            isphonehasError = true;
                        } else {
                            tvErrorPhone.setVisibility(View.GONE);
                            isphonehasError = false;
                        }
                    }

                }
            });

            btn_cancel_pp = v.findViewById(R.id.btn_cancel_pp);
            btn_save_pp = v.findViewById(R.id.btn_save_pp);
            setData();
            if ((memberProfileModel != null) && (Constants.PpView_Type.equals("update"))) {
                mViewModel.setData(getResources().getString(R.string.practice_partners));
                tv_add_pp.setText(R.string.edit_practice_partners);
                et_Fname.setText(memberProfileModel.getFirst_name());

                et_designation.setText(memberProfileModel.getDesignation());
                et_specialist.setText(memberProfileModel.getPractice());
                et_email.setText(memberProfileModel.getEmail());
                et_phonenumber.setText(memberProfileModel.getPhone());
            } else {
                mViewModel.setData(getResources().getString(R.string.add_practice_partners));
                tv_add_pp.setText(R.string.add_practice_partners);
                et_Fname.setText("");

                et_designation.setText("");
                et_specialist.setText("");
                et_email.setText("");
                et_phonenumber.setText("");
            }

            btn_save_pp.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if ((Objects.requireNonNull(et_Fname.getText()).toString().isEmpty())  || (Objects.requireNonNull(et_designation.getText()).toString().isEmpty()) || (Objects.requireNonNull(et_specialist.getText()).toString().isEmpty()) || (Objects.requireNonNull(et_email.getText()).toString().isEmpty()) || (Objects.requireNonNull(et_phonenumber.getText()).toString().isEmpty()) || (isphonehasError) || (isemailhasError)) {
                        String msg = "Please enter the";
                        if (Objects.requireNonNull(et_Fname.getText()).toString().isEmpty()) {
                            msg = msg + " First Name";
                        }
//                        if (Objects.requireNonNull(et_Lname.getText()).toString().isEmpty()) {
//                            if (msg.equals("Please enter the")) {
//                                msg = msg + " Last Name";
//                            } else {
//                                msg = msg + ", Last Name";
//                            }
//                        }
                        if (Objects.requireNonNull(et_designation.getText()).toString().isEmpty()) {
                            if (msg.equals("Please enter the")) {
                                msg = msg + " Designation";
                            } else {
                                msg = msg + ", Designation";
                            }
                        }
                        if (Objects.requireNonNull(et_specialist.getText()).toString().isEmpty()) {
                            if (msg.equals("Please enter the")) {
                                msg = msg + " Specialist";
                            } else {
                                msg = msg + ", Specialist";
                            }
                        }
                        if (Objects.requireNonNull(et_email.getText()).toString().isEmpty()) {
                            if (msg.equals("Please enter the")) {
                                msg = msg + " Email";
                            } else {
                                msg = msg + ", Email";
                            }
                        } else {
                            if (isemailhasError) {
                                if (msg.equals("Please enter the")) {
                                    msg = msg + " valid Email";
                                } else {
                                    msg = msg + ", valid Email";
                                }
                            }
                        }
                        if (Objects.requireNonNull(et_phonenumber.getText()).toString().isEmpty()) {
                            if (msg.equals("Please enter the")) {
                                msg = msg + " Phone";
                            } else {
                                msg = msg + ", Phone";
                            }
                        } else {
                            if (isphonehasError) {
                                if (msg.equals("Please enter the")) {
                                    msg = msg + " valid Phone number";
                                } else {
                                    msg = msg + ", valid Phone number";
                                }
                            }
                        }
                        AndroidUtils.showAlert(msg, getActivity());
                    } else {
                        if ((memberProfileModel != null) && (Constants.PpView_Type.equals("update"))) {
                            SaveDetailsPp("update");
                        } else {
                            SaveDetailsPp("add");
                            Constants.PpView_Type = "";
                        }
                    }
                }
            });
            btn_cancel_pp.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    clearDetails();
                }
            });
            iv_cancel.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    clearDetails();
                }
            });
        } catch (Resources.NotFoundException e) {
            e.fillInStackTrace();
        }
    }

    private void SaveDetailsPp(String save_type) {
//        https:api.staging.digicoffer.com/professional/partners/add
        progress_dialog = AndroidUtils.get_progress(getActivity());
        try {
            JSONObject jsonObject = new JSONObject();
//        {"first_name":"bharathi","last_name":"ganesh","designation":"Developer","phone":"1111111111","email":"bharathi.ganesh@digicoffer.com","practice":"Swift, Java"}
            jsonObject.put("name", Objects.requireNonNull(et_Fname.getText()).toString());

            jsonObject.put("designation", Objects.requireNonNull(et_designation.getText()).toString());
            jsonObject.put("practice", Objects.requireNonNull(et_specialist.getText()).toString());
            jsonObject.put("email", Objects.requireNonNull(et_email.getText()).toString());
            jsonObject.put("phone", Objects.requireNonNull(et_phonenumber.getText()).toString());

            if (save_type.equals("update"))
                WebServiceHelper.callHttpWebService(this, getContext(), WebServiceHelper.RestMethodType.PUT, "v3/practice-partner/" + memberProfileModel.getId(), "Update_Pp", jsonObject.toString());
            else
                WebServiceHelper.callHttpWebService(this, getContext(), WebServiceHelper.RestMethodType.POST, "v3/practice-partner", "Add_Pp", jsonObject.toString());

        } catch (Exception e) {
            e.fillInStackTrace();
        }
    }

    private void setData() {
        tv_add_pp.setText(R.string.add_practice_partners);
        tv_add_pp.setTextSize(DynamicUtils.twenty);
        tv_Fname.setText(R.string.name);
        tv_Lname.setText(R.string.last_name);
        tv_designation.setText(R.string.desg);
        tv_specialist.setText(R.string.specs);
        tv_email.setText(R.string.email_add);
        tv_phonenumber.setText(R.string.phone_no);

        et_Fname.setHint(R.string.name);
        et_Lname.setHint(R.string.last_name);
        et_designation.setHint(R.string.desg);
        et_specialist.setHint(R.string.specs);
        et_email.setHint(R.string.email_add);
        et_phonenumber.setHint(R.string.phone_no);
    }

    private void clearDetails() {
        et_Fname.setText("");
        et_Lname.setText("");
        et_designation.setText("");
        et_specialist.setText("");
        et_email.setText("");
        et_phonenumber.setText("");
        Constants.Edit_OR_View = "View";
        firmProfile.ChangeBackGround();
        mViewModel.setData(getResources().getString(R.string.practice_partners));
        firmProfile.NavFragment(PracticePartnerView.newInstance(firmProfile));
    }


    @Override
    public void onAsyncTaskComplete(HttpResultDo httpResult) {
        // Dismiss progress dialog
        if (progress_dialog != null && progress_dialog.isShowing()) {
            AndroidUtils.dismiss_dialog(progress_dialog);
        }

        String responseContent = httpResult.getResponseContent(); // raw response
        int statusCode = httpResult.getStatus_code(); // HTTP status
        String requestType = httpResult.getRequestType();

        Log.d("Request_Type", requestType);

        try {
            JSONObject resultJson = new JSONObject(responseContent);

            // Handle Add_Pp request
            if (Objects.equals(requestType, "Add_Pp")) {
                // Handle Unauthorized explicitly
                if (statusCode == 401) {
                    String msg = resultJson.optString("msg", "Unauthorized access");
                    AndroidUtils.showAlert(msg, getActivity());
                    return;
                }

                boolean isError = resultJson.optBoolean("error", false);
                String msg = resultJson.optString("msg", isError ? "Something went wrong" : "Operation completed");
                AndroidUtils.showAlert(msg, getActivity());

                if (!isError) {
                    clearDetails();
                }
            }
            // Handle Update_Pp request
            else if (Objects.equals(requestType, "Update_Pp")) {
                boolean isError = resultJson.optBoolean("error", false);
                String msg = resultJson.optString("msg", isError ? "Something went wrong" : "Operation completed");
                AndroidUtils.showAlert(msg, getActivity());

                if (!isError) {
                    clearDetails();
                }
            }
            // Optional: handle other request types here
            else {
                // Default handling for other requests
                boolean isError = resultJson.optBoolean("error", false);
                String msg = resultJson.optString("msg", isError ? "Something went wrong" : "Operation completed");
                AndroidUtils.showAlert(msg, getActivity());
            }

        } catch (JSONException e) {
            e.printStackTrace();
            AndroidUtils.showAlert("Something went wrong", getActivity());
        } catch (Exception e) {
            e.printStackTrace();
            AndroidUtils.showAlert("An unexpected error occurred", getActivity());
        }
    }

}

File: app/src/main/java/com/digicoffer/lauditor/FirmProfile/PracticePartnerAdd.kt
package com.digicoffer.lauditor.FirmProfile

import android.app.Dialog
import android.os.Bundle
import android.text.Editable
import android.text.InputType
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.TextView
import androidx.activity.OnBackPressedCallback
import androidx.appcompat.widget.LinearLayoutCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.Constants
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.DynamicUtils
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.NewModel
import com.digicoffer.lauditor.CommonFiles.ValidationUtils.Validation
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.Webservice.AsyncTaskCompleteListener
import com.digicoffer.lauditor.Webservice.CommonApiHelper.WebServiceHelper
import com.digicoffer.lauditor.Webservice.HttpResultDo
import com.google.android.material.textfield.TextInputEditText
import org.json.JSONException
import org.json.JSONObject

class PracticePartnerAdd(
    private var memberProfileModel: MemberProfileModel?,
    private val firmProfile: FirmProfile
) : Fragment(), AsyncTaskCompleteListener, View.OnClickListener {

    private var tv_add_pp: TextView? = null
    private var tv_Fname: TextView? = null
    private var tv_Lname: TextView? = null
    private var tv_designation: TextView? = null
    private var tv_specialist: TextView? = null
    private var tv_email: TextView? = null
    private var tv_phonenumber: TextView? = null

    private var et_Fname: TextInputEditText? = null
    private var et_Lname: TextInputEditText? = null
    private var et_designation: TextInputEditText? = null
    private var et_specialist: TextInputEditText? = null
    private var et_email: TextInputEditText? = null
    private var et_phonenumber: TextInputEditText? = null

    private var btn_cancel_pp: Button? = null
    private var btn_save_pp: Button? = null
    private var progress_dialog: Dialog? = null
    private var isemailhasError = false
    private var iv_cancel: ImageView? = null
    private var isphonehasError = false
    private var response_email: TextView? = null
    private var tvErrorPhone: TextView? = null
    private var last_name_layout: LinearLayoutCompat? = null
    private var mViewModel: NewModel? = null

    override fun onClick(view: View) {}

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.activity_practice_partner_add, container, false)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        setupOnBackPressed()
        super.onCreate(savedInstanceState)
    }

    private fun setupOnBackPressed() {
        requireActivity().onBackPressedDispatcher.addCallback(
            this,
            object : OnBackPressedCallback(true) {
                override fun handleOnBackPressed() {
                    if (isEnabled) {
                        isEnabled = false
                        requireActivity().onBackPressed()
                    }
                }
            })
    }

    override fun onViewCreated(v: View, savedInstanceState: Bundle?) {
        super.onViewCreated(v, savedInstanceState)
        try {
            mViewModel = ViewModelProvider(requireActivity()).get(NewModel::class.java)
            tv_add_pp = v.findViewById(R.id.tv_add_pp)
            tv_Fname = v.findViewById(R.id.tv_Fname)
            tv_Lname = v.findViewById(R.id.tv_Lname)
            iv_cancel = v.findViewById(R.id.iv_cancel)
            tv_designation = v.findViewById(R.id.tv_designation)
            tv_specialist = v.findViewById(R.id.tv_specialist)
            tv_email = v.findViewById(R.id.tv_email)
            tv_phonenumber = v.findViewById(R.id.tv_phonenumber)
            last_name_layout = v.findViewById(R.id.last_name_layout)
            last_name_layout?.visibility = View.GONE

            et_Fname = v.findViewById(R.id.et_Fname)
            et_Fname?.addTextChangedListener(Validation(et_Fname))

            et_Lname = v.findViewById(R.id.et_Lname)
            et_Lname?.visibility = View.GONE

            et_designation = v.findViewById(R.id.et_designation)
            et_designation?.addTextChangedListener(Validation(et_designation))

            et_specialist = v.findViewById(R.id.et_specialist)
            et_specialist?.addTextChangedListener(Validation(et_specialist))

            response_email = v.findViewById(R.id.response_email)
            tvErrorPhone = v.findViewById(R.id.tvErrorPhone)

            et_email = v.findViewById(R.id.et_email)
            et_email?.addTextChangedListener(Validation(et_email))
            et_email?.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                    response_email?.visibility = View.GONE
                    isemailhasError = false
                }

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

                override fun afterTextChanged(s: Editable?) {
                    val emailText = et_email?.text?.toString()?.trim() ?: ""
                    if (emailText.isNotEmpty()) {
                        if (!AndroidUtils.isValidEmail(emailText)) {
                            response_email?.visibility = View.VISIBLE
                            response_email?.setText(R.string.enter_a_valid_email_address)
                            isemailhasError = true
                        } else {
                            response_email?.visibility = View.GONE
                            isemailhasError = false
                        }
                    }
                }
            })

            et_phonenumber = v.findViewById(R.id.et_phonenumber)
            et_phonenumber?.inputType = InputType.TYPE_CLASS_NUMBER
            et_phonenumber?.addTextChangedListener(Validation(et_phonenumber))
            AndroidUtils.NumberFilter(et_phonenumber, true)

            et_phonenumber?.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                    tvErrorPhone?.visibility = View.GONE
                    isphonehasError = false
                }

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

                override fun afterTextChanged(s: Editable?) {
                    val text = s?.toString() ?: ""
                    if (text.isNotEmpty()) {
                        if (text.length < 10) {
                            tvErrorPhone?.visibility = View.VISIBLE
                            tvErrorPhone?.setText(R.string.please_enter_the_10_digit_mobile_number)
                            isphonehasError = true
                        } else {
                            tvErrorPhone?.visibility = View.GONE
                            isphonehasError = false
                        }
                    }
                }
            })

            btn_cancel_pp = v.findViewById(R.id.btn_cancel_pp)
            btn_save_pp = v.findViewById(R.id.btn_save_pp)

            setData()

            val model = memberProfileModel
            if (model != null && Constants.PpView_Type == "update") {
                mViewModel?.setData(resources.getString(R.string.practice_partners))
                tv_add_pp?.setText(R.string.edit_practice_partners)
                et_Fname?.setText(model.first_name)
                et_designation?.setText(model.designation)
                et_specialist?.setText(model.practice)
                et_email?.setText(model.email)
                et_phonenumber?.setText(model.phone)
            } else {
                mViewModel?.setData(resources.getString(R.string.add_practice_partners))
                tv_add_pp?.setText(R.string.add_practice_partners)
                et_Fname?.setText("")
                et_designation?.setText("")
                et_specialist?.setText("")
                et_email?.setText("")
                et_phonenumber?.setText("")
            }

            btn_save_pp?.setOnClickListener {
                val fName = et_Fname?.text?.toString() ?: ""
                val desg = et_designation?.text?.toString() ?: ""
                val spec = et_specialist?.text?.toString() ?: ""
                val email = et_email?.text?.toString() ?: ""
                val phone = et_phonenumber?.text?.toString() ?: ""

                if (fName.isEmpty() || desg.isEmpty() || spec.isEmpty() || email.isEmpty() || phone.isEmpty() || isphonehasError || isemailhasError) {
                    var msg = "Please enter the"
                    if (fName.isEmpty()) {
                        msg += " First Name"
                    }
                    if (desg.isEmpty()) {
                        msg += if (msg == "Please enter the") " Designation" else ", Designation"
                    }
                    if (spec.isEmpty()) {
                        msg += if (msg == "Please enter the") " Specialist" else ", Specialist"
                    }
                    if (email.isEmpty()) {
                        msg += if (msg == "Please enter the") " Email" else ", Email"
                    } else if (isemailhasError) {
                        msg += if (msg == "Please enter the") " valid Email" else ", valid Email"
                    }
                    if (phone.isEmpty()) {
                        msg += if (msg == "Please enter the") " Phone" else ", Phone"
                    } else if (isphonehasError) {
                        msg += if (msg == "Please enter the") " valid Phone number" else ", valid Phone number"
                    }
                    AndroidUtils.showAlert(msg, activity)
                } else {
                    if (memberProfileModel != null && Constants.PpView_Type == "update") {
                        SaveDetailsPp("update")
                    } else {
                        SaveDetailsPp("add")
                        Constants.PpView_Type = ""
                    }
                }
            }

            btn_cancel_pp?.setOnClickListener { clearDetails() }
            iv_cancel?.setOnClickListener { clearDetails() }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun SaveDetailsPp(save_type: String) {
        progress_dialog = AndroidUtils.get_progress(activity)
        try {
            val jsonObject = JSONObject()
            jsonObject.put("name", et_Fname?.text?.toString() ?: "")
            jsonObject.put("designation", et_designation?.text?.toString() ?: "")
            jsonObject.put("practice", et_specialist?.text?.toString() ?: "")
            jsonObject.put("email", et_email?.text?.toString() ?: "")
            jsonObject.put("phone", et_phonenumber?.text?.toString() ?: "")

            val model = memberProfileModel
            if (save_type == "update" && model != null) {
                WebServiceHelper.callHttpWebService(
                    this,
                    requireContext(),
                    WebServiceHelper.RestMethodType.PUT,
                    "v3/practice-partner/${model.id}",
                    "Update_Pp",
                    jsonObject.toString()
                )
            } else {
                WebServiceHelper.callHttpWebService(
                    this,
                    requireContext(),
                    WebServiceHelper.RestMethodType.POST,
                    "v3/practice-partner",
                    "Add_Pp",
                    jsonObject.toString()
                )
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun setData() {
        tv_add_pp?.setText(R.string.add_practice_partners)
        tv_add_pp?.setTextSize(android.util.TypedValue.COMPLEX_UNIT_SP, DynamicUtils.twenty.toFloat())
        tv_Fname?.setText(R.string.name)
        tv_Lname?.setText(R.string.last_name)
        tv_designation?.setText(R.string.desg)
        tv_specialist?.setText(R.string.specs)
        tv_email?.setText(R.string.email_add)
        tv_phonenumber?.setText(R.string.phone_no)

        et_Fname?.setHint(R.string.name)
        et_Lname?.setHint(R.string.last_name)
        et_designation?.setHint(R.string.desg)
        et_specialist?.setHint(R.string.specs)
        et_email?.setHint(R.string.email_add)
        et_phonenumber?.setHint(R.string.phone_no)
    }

    private fun clearDetails() {
        et_Fname?.setText("")
        et_Lname?.setText("")
        et_designation?.setText("")
        et_specialist?.setText("")
        et_email?.setText("")
        et_phonenumber?.setText("")
        Constants.Edit_OR_View = "View"
        firmProfile.ChangeBackGround()
        mViewModel?.setData(resources.getString(R.string.practice_partners))
        firmProfile.NavFragment(PracticePartnerView.newInstance(firmProfile))
    }

    override fun onAsyncTaskComplete(httpResult: HttpResultDo) {
        if (progress_dialog != null && progress_dialog!!.isShowing) {
            AndroidUtils.dismiss_dialog(progress_dialog)
        }

        val responseContent = httpResult.responseContent ?: ""
        val statusCode = httpResult.status_code
        val requestType = httpResult.requestType ?: ""

        Log.d("Request_Type", requestType)

        try {
            val resultJson = JSONObject(responseContent)

            if (requestType == "Add_Pp") {
                if (statusCode == 401) {
                    val msg = resultJson.optString("msg", "Unauthorized access")
                    AndroidUtils.showAlert(msg, activity)
                    return
                }

                val isError = resultJson.optBoolean("error", false)
                val msg = resultJson.optString(
                    "msg",
                    if (isError) "Something went wrong" else "Operation completed"
                )
                AndroidUtils.showAlert(msg, activity)

                if (!isError) {
                    clearDetails()
                }
            } else if (requestType == "Update_Pp") {
                val isError = resultJson.optBoolean("error", false)
                val msg = resultJson.optString(
                    "msg",
                    if (isError) "Something went wrong" else "Operation completed"
                )
                AndroidUtils.showAlert(msg, activity)

                if (!isError) {
                    clearDetails()
                }
            } else {
                val isError = resultJson.optBoolean("error", false)
                val msg = resultJson.optString(
                    "msg",
                    if (isError) "Something went wrong" else "Operation completed"
                )
                AndroidUtils.showAlert(msg, activity)
            }
        } catch (e: JSONException) {
            e.printStackTrace()
            AndroidUtils.showAlert("Something went wrong", activity)
        } catch (e: Exception) {
            e.printStackTrace()
            AndroidUtils.showAlert("An unexpected error occurred", activity)
        }
    }
}
